lib/rules/unit-no-unknown/index.js

```diff
@@ -48,122 +58,149 @@ const rule = (primary, secondaryOptions) => {
-						node.name === 'media' &&
-						node.params.toLowerCase().includes('resolution')
-					) {
-						let ignoreUnit = false;
-
-						mediaParser(node.params).walk((mediaNode, _i, mediaNodes) => {
-							const lastMediaNode = mediaNodes[mediaNodes.length - 1];
-
-							if (
-								mediaNode.value.toLowerCase().includes('resolution') &&
-								lastMediaNode &&
-								lastMediaNode.sourceIndex === valueNode.sourceIndex
-							) {
-								ignoreUnit = true;
-
-								return false;
-							}
-						});
-
-						if (ignoreUnit) {
-							return;
-						}
-					}
+				return;
+			}
 
-					if (node.type === 'decl') {
-						if (node.prop.toLowerCase() === 'image-resolution') {
-							return;
-						}
+			const [tokenType, , , endIndex, tokenValue] = componentValue.value;
 
-						if (/^(?:-webkit-)?image-set[\s(]/i.test(value)) {
-							const imageSet = parsedValue.nodes.find(
-								(n) => vendor.unprefixed(n.value) === 'image-set',
-							);
+			if (tokenType !== TokenType.Dimension) return;
 
-							assert(imageSet);
-							assert('nodes' in imageSet);
-							const imageSetLastNode = imageSet.nodes[imageSet.nodes.length - 1];
+			if (optionsMatches(secondaryOptions, 'ignoreUnits', tokenValue.unit)) return;
 
-							assert(imageSetLastNode);
-							const imageSetValueLastIndex = imageSetLastNode.sourceIndex;
+			const unit = tokenValue.unit.toLowerCase();
 
-							if (imageSetValueLastIndex >= valueNode.sourceIndex) {
-								return;
-							}
-						}
-					}
-				}
+			if (unit === 'x' && state.allowX) return;
 
-				const index = getIndex(node);
+			if (units.has(unit) && unit !== 'x') return;
 
-				report({
-					index: index + valueNode.sourceIndex + number.length,
-					endIndex: index + valueNode.sourceEndIndex,
-					message: messages.rejected,
-					messageArgs: [unit],
-					node,
-					result,
-					ruleName,
-				});
+			const startIndex = getIndex(node) + (endIndex + 1) - unit.length;
+
+			report({
+				message: messages.rejected,
+				messageArgs: [tokenValue.unit],
+				node,
+				index: startIndex,
+				endIndex: startIndex + unit.length,
+				result,
+				ruleName,
 			});
-		}
+		};
 
 		root.walkAtRules(/^media$/i, (atRule) => {
-			if (!isStandardSyntaxAtRule(atRule)) {
-				return;
-			}
+			if (!isStandardSyntaxAtRule(atRule)) return;
+
+			const params = getAtRuleParams(atRule);
+
+			const tokens = tokenizeIfValueMightContainUnknownUnits(params);
+
+			if (!tokens) return;
 
-			check(atRule, atRule.params, atRuleParamIndex);
+			parseFromTokens(tokens).forEach((mediaQuery) => {
+				const state = {
+					ignored: false,
+					allowX: false,
+				};
+
+				mediaQuery.walk((entry) => {
+					if (entry.state?.ignored) return;
+
+					if (isMediaFeature(entry.node)) {
+						const name = entry.node.getName();
+
+						if (entry.state && RESOLUTION_MEDIA_FEATURE_NAME.test(name)) {
+							entry.state.allowX = true;
+						}
+					}
+
+					if (isFunctionNode(entry.node) || isTokenNode(entry.node)) {
```

Comment:
[ask] Does converting to `else if` make sense to reduce extra evaluations?

```suggestion
					} else if (isFunctionNode(entry.node) || isTokenNode(entry.node)) {
```

Reply:
yes! thanks 🙇